syft/pkg/cataloger/common/cpe/generate.go

```diff
@@ -8,18 +8,23 @@ import (
 	"strings"
 
 	"github.com/anchore/syft/internal"
+	"github.com/anchore/syft/internal/log"
 	"github.com/anchore/syft/syft/pkg"
 	"github.com/facebookincubator/nvdtools/wfn"
 )
 
-func newCPE(product, vendor, version, targetSW string) wfn.Attributes {
+func newCPE(product, vendor, version, targetSW string) *wfn.Attributes {
 	cpe := *(wfn.NewAttributesWithAny())
 	cpe.Part = "a"
 	cpe.Product = product
 	cpe.Vendor = vendor
 	cpe.Version = version
 	cpe.TargetSW = targetSW
-	return cpe
+	if pkg.ValidateCPEString(pkg.CPEString(cpe)) != nil {
+		log.Warnf("could not generate CPE for %+v", cpe)
```

Comment:
Should this be a warning? Probably `Debugf` since we'll be using this function in multiple places from multiple SBOM inputs

Reply:
Here is a screenshot what the output looks like with the warning. I lean slightly towards keeping this as a warning since I think this is a category of alerting the user that something went wrong vs a procedural step like `cataloging with "ruby-gemspec-cataloger"`

<img width="1792" alt="Screen Shot 2022-02-25 at 2 19 02 PM" src="https://user-images.githubusercontent.com/32073428/155774894-6ff66055-2f3b-4386-931a-0455129e390d.png">

Reply:
My concern is more... let's say we import SPDX or CycloneDX that has thousands of components and thousands of CPEs, we'll see this error a bunch of the tool used to generate said document isn't doing a good job, even though we might be able to still provide some useful things like Grype vuln matching with generated CPEs

Reply:
... which leans to my second comment, can't we just encode these somehow? https://github.com/anchore/syft/pull/850#discussion_r815033319

Reply:
I could go either way on this (leave it or take it out). Taking it out is alright since its warning about synthetic data problems and not from data we found in the source scanned.